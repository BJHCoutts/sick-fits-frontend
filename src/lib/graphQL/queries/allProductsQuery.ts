import { gql } from "@apollo/client"

export const ALL_PRODUCTS_QUERY = gql`
query ALL_PRODUCTS_QUERY{
  allProducts {
    id
    name
    description
    photo {
      id
      image {
        id
        publicUrlTransformed
      }
      altText
    }
    status
  }
}
`