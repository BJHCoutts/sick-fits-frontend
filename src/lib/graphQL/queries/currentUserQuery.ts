import { gql } from "@apollo/client"

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            name
            description
            price
            photo {
              id
              altText
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`