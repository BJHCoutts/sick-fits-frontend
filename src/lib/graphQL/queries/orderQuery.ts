import { gql } from "@apollo/client"

export const ORDER_QUERY = gql`
  query ORDER_QUERY($id: ID!) {
    order: Order(where: { id: $id }) {
      id
      charge
      total
      user {
        id
      }
      items {
        id
        name
        description
        price
        quantity
        photo {
          image {
            publicUrlTransformed
          }
        }
      }
    }
  }
`