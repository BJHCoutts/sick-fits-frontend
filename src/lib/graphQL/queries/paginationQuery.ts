import { gql } from "@apollo/client"

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY{
		_allProductsMeta {
			count
		}
	}
`