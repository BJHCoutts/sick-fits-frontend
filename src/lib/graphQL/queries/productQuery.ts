import { gql } from "@apollo/client"

export const PRODUCT_QUERY = gql`
	query PRODUCT_QUERY (
		$id: ID!
		){
		Product (where: {id: $id}) {
			id
			name
			description
			photo {
				id
				image {
					id
					publicUrlTransformed
				}
				altText
			}
			status
			price
		}
	}
`