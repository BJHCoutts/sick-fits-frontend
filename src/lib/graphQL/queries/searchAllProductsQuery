import { gql } from "@apollo/client"

export const searchAllProductsQuery = gql`
	query SEARCH_ALL_PRODUCTS_QUERY($searchTerm:String!) {
		allProducts (where: { 
			OR:[
				{name_contains_i: $searchTerm}, 
				{  description_contains_i: $searchTerm}
			]
		})
		{
			id
			name
			description
			photo {
				id
				image {
					id
					publicUrlTransformed
				}
				altText
			}
		}
	}
`