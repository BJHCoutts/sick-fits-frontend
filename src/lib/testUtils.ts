import casual from 'casual'
import { PAGINATION_QUERY } from './graphQL/queries/paginationQuery'
import { TProduct } from './types/TProduct'
import { TCartItem } from './types/TCartItem'
import { TUser } from './types/TUser'

// seed it so we get consistent results
casual.seed(777)

export const fakeItem = (): TProduct => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	photo: {
		id: 'abc123',
		altText: 'dogs are best',
		image: {
			id: 'abc123',
			publicUrlTransformed: 'dog.jpg',
		},
	},
	name: 'dogs are best',
	description: 'dogs',
	status: 'AVAILABLE',
})

export const fakeUser = (overrides: {}): TUser => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...overrides,
})

// const fakeOrderItem = () => ({
//   __typename: 'OrderItem',
//   id: casual.uuid,
//   image: {
//     image: `${casual.word}.jpg`,
//   },
//   name: casual.words(),
//   price: 4234,
//   quantity: 1,
//   description: casual.words(),
// });

// const fakeOrder = () => ({
//   __typename: 'Order',
//   id: 'ord123',
//   charge: 'ch_123',
//   total: 40000,
//   items: [fakeOrderItem(), fakeOrderItem()],
//   createdAt: '2022-12-11T20:16:13.797Z',
//   user: fakeUser(),
// });

export const fakeCartItem = (overrides: {}): TCartItem => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 4,
	product: fakeItem(),
	user: fakeUser({}),
	...overrides,
})

// // Fake LocalStorage
// class LocalStorageMock {
//   constructor() {
//     this.store = {};
//   }

//   clear() {
//     this.store = {};
//   }

//   getItem(key) {
//     return this.store[key] || null;
//   }

//   setItem(key, value) {
//     this.store[key] = value.toString();
//   }

//   removeItem(key) {
//     delete this.store[key];
//   }
// }

export function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	]
}

// export
// makePaginationMocksFor,
// LocalStorageMock,
// fakeItem,
// fakeUser,
// fakeCartItem,
// fakeOrder,
// fakeOrderItem,
