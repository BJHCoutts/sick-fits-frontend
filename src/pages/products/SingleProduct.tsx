import { useQuery } from '@apollo/client'
import { PRODUCT_QUERY } from '../../lib/graphQL/queries/productQuery'
import DisplayError from '../../components/ErrorMessage'
import { TProduct } from '../../lib/types/TProduct'
import Head from 'next/head'
import styled from 'styled-components'

export default function SingleProduct({
	id,
}: {
	id: string | string[] | undefined
}) {
	const { data, error, loading } = useQuery(PRODUCT_QUERY, {
		variables: { id },
	})

	if (loading) return <p>Loading...</p>
	if (error) return <DisplayError error={error} />

	const { Product }: { Product: TProduct } = data

	return (
		<>
			<Head>
				<title>{`Sick Fits | ${Product.name}`}</title>
			</Head>
			<SProduct data-testid="single-product">
				<img
					src={Product?.photo?.image.publicUrlTransformed}
					alt={Product?.photo?.altText}
				/>

				{!Product.photo && <h3 className="no-photo">No Photo</h3>}
				<div className="details">
					<h2>{Product.name}</h2>
					<p>{Product.description}</p>
				</div>
			</SProduct>
		</>
	)
}

const SProduct = styled.section`
	display: grid;
	grid-auto-columns: 1fr;
	grid-auto-flow: column;
	min-height: 800px;
	max-width: var(--max-width);
	align-items: top;
	gap: 2rem;
	img {
		object-fit: contain;
	}
`
