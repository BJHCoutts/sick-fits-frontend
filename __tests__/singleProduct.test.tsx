import { render, screen } from '@testing-library/react'
import { PRODUCT_QUERY } from '../src/lib/graphQL/queries/productQuery'
import { fakeProduct } from '../src/lib/testUtils'
import { MockedProvider } from '@apollo/client/testing'
import SingleProduct from '../src/pages/products/SingleProduct'

describe('<SingleProduct />', () => {
	it('renders with correct data', async () => {
		const mocks = [
			{
				request: {
					query: PRODUCT_QUERY,
					variables: {
						id: '123',
					},
				},
				result: {
					data: {
						Product: fakeProduct,
					},
				},
			},
		]
		const { container, debug } = render(
			<MockedProvider mocks={mocks}>
				<SingleProduct id="123" />
			</MockedProvider>
		)
		debug()
		await screen.findByTestId('single-product')
		expect(container).toMatchSnapshot()
	})
})

it('Errors out when product is not found', async () => {
	const mockError = [
		{
			request: {
				query: PRODUCT_QUERY,
				variables: {
					id: '123',
				},
			},
			result: {
				errors: [{ message: 'Product not found' }],
			},
		},
	]

	const { container, debug } = render(
		<MockedProvider mocks={mockError}>
			<SingleProduct id="123" />
		</MockedProvider>
	)

	await screen.findByTestId('graphql-error')
	debug()
	expect(container).toHaveTextContent('Shoot!')
})
