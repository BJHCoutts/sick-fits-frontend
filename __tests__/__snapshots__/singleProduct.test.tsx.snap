// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> renders with correct data 1`] = `
<div>
  <section
    class="sc-gsTDqH llXjTR"
    data-testid="single-product"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </section>
</div>
`;
